### PR TITLE
driver/counter/counter_ll_stm32_rtc.c: Add 1 tick to alarm

### LINKED ISSUE #25366: tests/drivers/counter/counter_basic_api: instable test status on STM32 boards
Warning: it is needed to cherry picking PR #25364 to be able to reach this stage.

While exectuing test **tests/drivers/counter/counter_basic_api**, from time to time, test fails during 
step **test_short_relative_alarm**,  with 1 less callback count than expected.

**To Reproduce**
1. west -v build -p auto  -b nucleo_f746zg tests/drivers/counter/counter_basic_api
2. See error

**Expected behavior**
Test Passed

**Screenshots or console output**

> starting test - test_short_relative_alarm
> D: Set Alarm: 10
> 
> Testing RTC_0
> D: Set Alarm: 20
> 
> D: Set Alarm: 23
> 
> D: Set Alarm: 26
> 
> D: Set Alarm: 29
> 
> D: Set Alarm: 32
> 
> D: Set Alarm: 34
> 
> D: Set Alarm: 37
> 
> D: Set Alarm: 40
> 
> D: Set Alarm: 43
> 
> D: Set Alarm: 46
> 
> D: Set Alarm: 49
> 
> D: Set Alarm: 52
> 
> D: Set Alarm: 55
> 
> D: Set Alarm: 58
> 
> D: Set Alarm: 60
> 
> D: Set Alarm: 63
> 
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:711: test_short_relative_alarm_instance: (i + 1 equal to alarm_cnt)
> RTC_0: Expected 17 callbacks, got 16

### PR DESCRIPTION
**Pull request by ABOSTM**
Add +1 tick to alarm in order to compensate the partially started tick.
Alarm will expire between requested ticks and ticks+1.
In case only 1 tick is requested, it will avoid that +1 Tick event occurs
before alarm setting is finished.

Fixes #25366

### CODE REVIEW

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 1]
```diff
@@ -143,7 +143,11 @@ static int rtc_stm32_set_alarm(struct device *dev, u8_t chan_id,
 	data->user_data = alarm_cfg->user_data;
 
 	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
-		ticks += now;
+		/* Add +1 in order to compensate the partially started tick. */
+		/* Alarm will expire between requested ticks and ticks+1. */
+		/* In case only 1 tick is requested, it will avoid */
+		/* that tick+1 event occurs before alarm setting is finished. */
```
Comment (gmarull) [inline drivers/counter/counter_ll_stm32_rtc.c:149]: Use a continuous comment,

```c
/* a long
 * comment
 */
```
I guess that by doing this adjustment alarm will always be triggered in `ticks` plus the elapsed amount of the potentially started tick. So in the worst case, user will experience 1 tick more and in the best case the exact amount of ticks. If this adjustment is not done, user will experience in the worst case 1 tick less and again the exact amount of ticks for best case. Assuming setting takes < 1s, the 1s case could be treated differently so other cases are not affected, right?

Comment (ABOSTM) [reply drivers/counter/counter_ll_stm32_rtc.c]: > Use a continuous comment

Done

Yes it is possible to handle the special case of 1tick request and let other case unmodified.
But if I add +1 tick only when 1 tick is requested, it means that requesting 1Tick will be the same than requesting 2 ticks.
Also I think normal (more intuitive) that when application request N ticks we guaranntee at least N ticks and not less.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 2]
```diff
@@ -143,7 +143,12 @@ static int rtc_stm32_set_alarm(struct device *dev, u8_t chan_id,
 	data->user_data = alarm_cfg->user_data;
 
 	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
-		ticks += now;
+		/* Add +1 in order to compensate the partially started tick.
+		 * Alarm will expire between requested ticks and ticks+1.
+		 * In case only 1 tick is requested, it will avoid */
```
Comment (gmarull) [inline drivers/counter/counter_ll_stm32_rtc.c:148]: remove `*/`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 1 above)
Comment (gmarull) [reply drivers/counter/counter_ll_stm32_rtc.c]: Ok, fair point.

**Comment by gmarull** [comment]
@ABOSTM I think PR number needs to be removed from commit message. Also, you can cherry-pick the commit in here https://github.com/zephyrproject-rtos/zephyr/pull/25371

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 2 above)
Comment (ABOSTM) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done, Sorry for the inconvenience.
Thanks for the hint, unfortunately checkpatch is not supported on Windows.

**Comment by ABOSTM** [comment]
@gmarull,

> @ABOSTM I think PR number needs to be removed from commit message. Also, you can cherry-pick the commit in here #25371

I am not sure to catch your point: there is no PR number is the commit message of this RP.
So I guess your comment apllies to  #25371 instead.
Also do you suggest to make only 1 PR for both commit from #25367 and #25371  ?

**Comment by gmarull** [comment]
@ABOSTM yes, I meant the ` PR` bit in #25371, sorry. It's ok to have 2 PRs, but as they are related they could have gone in a single PR and 2 commits.

**Review by pabigot — APPROVED** [review]
I think this is right: a small offset of 1 or 2 ticks depending on hardware is generally added to ensure late-to-set alarms don't occur.

**Review by gmarull — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**